<font size="+3"><b>Assignment 2: Linear Models and Validation Metrics</b></font>

***
* **Full Name** = Justin Pham
* **UCID** = 30139323
***

In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.
You can use the Table of Content on the left side of this notebook to efficiently navigate within this documents.

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|           **Part 2: Regression**           |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    0.5    |
| Step 3: Implement Machine Learning   Model |     1     |
|            Step 4: Validate Mode           |     1     |
|          Step 5: Visualize Results         |     1     |
|                  Questions                 |     2     |
|             Process Description            |     4     |
|  **Part 3:   Observations/Interpretation** |   **3**   |
|           **Part 4: Reflection**           |   **2**   |
|                  **Total**                 |   **30**  |
|                                            |           |
|                  **Bonus**                 |           |
|         **Part 5: Bonus Question**         |   **4**   |

# **Part 1: Classification (14.5 marks total)**

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|                  **Total**                 |  **14.5** |

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:


## **Step 0:** Import Libraries

In [1]:
import numpy as np
import pandas as pd

C:\Users\night\AppData\Local\Temp\ipykernel_20888\1662815981.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [2]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets.loaders import load_spam

# Load the spam dataset
data = load_spam(return_dataset=True)

# TO DO: Print size and type of X and y
X, y = data.to_data()
print(f"X.shape: {X.shape}, Y.shape: y.shape")
print(f"type(x): {type(X)} \ntypes(y): {type(y)}")

X.shape: (4600, 57), Y.shape: y.shape
type(x): <class 'pandas.core.frame.DataFrame'> 
types(y): <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [3]:
# TO DO: Check if there are any missing values and fill them in if necessary
checkX = [field for field in X.isna().sum() if field > 0]
print(checkX)
print(y.isna().sum())

[]
0


For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [4]:
# TO DO: Create X_small and y_small
from sklearn.model_selection import train_test_split
X_train_temp, X_small, y_train_temp, y_small = train_test_split(X, y, test_size=0.05, random_state=0)
# X_train_temp and y_train_temp are not used in this example

## **Step 3:** Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

In [5]:
from sklearn.linear_model import LogisticRegression

# TO DO: Create a Logistic Regression model
# Model uses X and y
modelXY = LogisticRegression(max_iter=2000)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
modelXY.fit(X_train, y_train)

# Model only uses the first two columns of X and y
modelXY2 = LogisticRegression(max_iter=2000)
X2_train, X2_test, y2_train, y2_test = train_test_split(X.iloc[:, :2], y, test_size=0.2, random_state=0)
modelXY2.fit(X2_train, y2_train)

# Model uses X_small and y_small
modelXYSmall = LogisticRegression(max_iter=2000)
X_small_train, X_small_test, y_small_train, y_small_test = train_test_split(X_small, y_small, test_size=0.05, random_state=0)
modelXYSmall.fit(X_small, y_small)


LogisticRegression(max_iter=2000)

## **Step 4:** Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

## **Step 5:** Visualize Results (4 marks for steps 3-5)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [6]:
from sklearn.metrics import accuracy_score, precision_score
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

results = pd.DataFrame(columns=["Data size", "training accuracy", "validation accuracy", "model"])

# Create a list of datasets and models
datasets = [
    (X_train, y_train, X_test, y_test, modelXY, "modelXY"),
    (X2_train, y2_train, X2_test, y2_test, modelXY2, "modelXY2"),
    (X_small_train, y_small_train, X_small_test, y_small_test, modelXYSmall, "modelXYSmall")
]

# Loop over the datasets and store the results in the results dataframe
for X_train, y_train, X_test, y_test, model, modelName in datasets:
    dataSize = len(X_train)
    trainingAccuracy = accuracy_score(y_train, model.predict(X_train))
    validationAccuracy = accuracy_score(y_test, model.predict(X_test))
    new_row = {"Data size": dataSize, "training accuracy": trainingAccuracy, "validation accuracy": validationAccuracy, "model": modelName}
    new_row_df = pd.DataFrame([new_row])
    results = pd.concat([results, new_row_df], ignore_index=True)

# Print the results
print(results)
print(X2_train.shape, X2_test.shape)
print(X_small_train.shape, X_small_test.shape)

  Data size  training accuracy  validation accuracy         model
0      3680           0.927717             0.938043       modelXY
1      3680           0.614946             0.593478      modelXY2
2       218           0.954128             0.750000  modelXYSmall
(3680, 2) (920, 2)
(218, 57) (12, 57)


C:\Users\night\AppData\Local\Temp\ipykernel_20888\1840995752.py:22: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  results = pd.concat([results, new_row_df], ignore_index=True)


## **Questions (4 marks)**
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

<font color='Green'><b>
1. From the data, we can see that if we supply the model with more information, the model is more likely to get more accurate predictions.
However, one thing to note is that while the model with only 2 columns, modelXY2, had much more rows (3680 vs 218 rows) than the model with only 5% of the data but all (57) columns, modelXYSmall, modelXYSmall performed much better in both training accuracy (0.61 vs 0.95) and validation accuracy(0.59 and 0.75) than modelXY2. This difference likely could have been due to the shape of the data. The greater the area (rows * columns), the more accurate the model may be.
One other thing to note is that modelXYSmall performed better than modelXY in training accuracy. This could be do to the model easily overfitting the smaller dataset, possibly memorizing it. However, since the dataset was much smaller, it could not predict unforeseen as well, evident in the much lower validation accuracy (0.75 and 0.92)

2. In the context of a classification problem, the concepts of false positives and false negatives are tied to the specific application:

    False Positive (FP): A false positive occurs when the model incorrectly predicts the positive class. For example, in a medical diagnosis scenario where the positive class is "disease present," a false positive would mean incorrectly diagnosing a healthy patient as sick.

    False Negative (FN): A false negative occurs when the model incorrectly predicts the negative class. Continuing with the medical example, a false negative would mean failing to diagnose a sick patient, incorrectly identifying them as healthy.

    However, in this case where we have data from a dataset classifying email as Spam or Non-Spam. According to the scikit documentation of the load_spam() dataset, The “spam” 
    concept is diverse: advertisements for products/web sites, make money fast schemes, chain letters, pornography… 
    With this definition, it is important that we do not mark important emails as spam as this could result in a user deleting this important email. A false positive in this case would be classifying an important email as spam where a false negative would be incorrectly marking a spam email as "not spam". Therefore, we can conclude that a false positive is much worse than a false negative in this case.

</b></font>

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>
1. Where did you source your code?
Code was written by Justin Pham, ChatGPT, scikit documentation, and some from ENSF444 Lab_3 validation_metrics. I often referenced the function names in Lab_3 including the use of train_test_split().
1. In what order did you complete the steps?
I completed the steps from top to bottom. 
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
I struggled with Part 1, Step 2: Data Processing, and asked ChatGPT the following prompt:
For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data. I did not need to modify the code other than add a comment explaining some of the code.
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
One of the challenges I faced during this lab was understanding what some of the questions were asking. They were not clear and simple enough for me to always understand. In the future labs, I am hoping to see an improvement in clarity.

</b></font>

# **Part 2: Regression (10.5 marks total)**

| **Question**                               | **Point** |
|--------------------------------------------|-----------|
| **Part 2: Regression**                     |           |
| Step 1: Data Input                         | 1         |
| Step 2: Data Processing                    | 0.5       |
| Step 3: Implement Machine Learning   Model | 1         |
| Step 4: Validate Mode                      | 1         |
| Step 5: Visualize Results                  | 1         |
| Questions                                  | 2         |
| Process Description                        | 4         |
| **Total**                                  | **10.5**  |

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [7]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets.loaders import load_concrete
# TO DO: Print size and type of X and y
data = load_concrete(return_dataset=True)
X, y = data.to_data()
print(f"X.shape: {X.shape}, Y.shape: y.shape")
print(f"type(x): {type(X)} \ntypes(y): {type(y)}")


X.shape: (1030, 8), Y.shape: y.shape
type(x): <class 'pandas.core.frame.DataFrame'> 
types(y): <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [8]:
# TO DO: Check if there are any missing values and fill them in if necessary
checkXNull = [field for field in X.isna().sum() if field > 0]
print(checkXNull)
print(y.isna().sum())

[]
0


## **Step 3:** Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [9]:
from sklearn.linear_model import LinearRegression
# TO DO: Create a Linear Regression model
model = LinearRegression()


In [10]:
from sklearn.model_selection import train_test_split
# Note: for any random state parameters, you can use random_state = 0
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

model.fit(X_train, y_train)

LinearRegression()

## **Step 4:** Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [11]:
from sklearn.metrics import mean_squared_error, r2_score
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean squared error: {mse}")
print(f"R2 score: {r2}")


Mean squared error: 95.63533482690428
R2 score: 0.6368981103411242


## **Step 5:** Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [12]:
import pandas as pd

# Create a pandas DataFrame 
results = pd.DataFrame(columns=["MSE", "R2"])

# Add data to the DataFrame
new_row = {"MSE": mse, "R2": r2}
new_row_df = pd.DataFrame([new_row])
results = pd.concat([results, new_row_df])

#  print the results
print(results)


         MSE        R2
0  95.635335  0.636898


C:\Users\night\AppData\Local\Temp\ipykernel_20888\1717950249.py:9: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  results = pd.concat([results, new_row_df])


## **Questions (2 marks)**
1. Did using a linear model produce good results for this dataset? Why or why not?

Given that the concrete dataset involves predicting concrete strength, a mean squared error of 95.635 and R2 of 0.63 suggests that the model has a moderate explanatory power. It indicates that the model captures a significant portion of the variance in concrete strength, but there's still room for improvement.

The MSE of 95.635 is moderate since according to the scikit load_concrete() documentation, the target's values range from 2.3 - 82.6.
The value of R2 ranges from 0 - 1 where the higher the value, the better the model understands the data.


## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:

<font color='Green'><b>
1. Where did you source your code?
Code was written by Justin Pham, scikit documentation, ENSF444 Lab_2 and some from ENSF444 Lab_3 validation_metrics. I often referenced the function names in Lab_2 and Lab_3 including the use of train_test_split().
1. In what order did you complete the steps?
I completed the steps from top to bottom. 
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
For Part 2, I did not need to use any generative AI.
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
Similar to the previous part, understanding what some of the questions were asking was challenging. They were not clear and simple enough for me to always understand. In the future labs, I am hoping to see an improvement in clarity. For example, one of the questions was if the linear model produced a "good result" on the dataset. Since this was a subjective question, I found it more difficult to answer.

</b></font>

# **Part 3: Observations/Interpretation (3 marks)**

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


<font color='Green'><b>
One notable pattern observed during this lab is the relationship between the size of the dataset and model accuracy. Generally, models trained with larger datasets exhibited higher accuracy. This trend was particularly evident in Part 1, Step 5: Visualize Results, where we analyzed the impact of data size on model performance.

In our examination, the validation accuracy of the Linear Regression model significantly decreased as the dataset size was reduced. Interestingly, when we compared the performance of a model trained on a very small dataset (modelXYSmall, with only 5% of the data) to one trained on a dataset with only two features (modelXY2), the former outperformed the latter despite its smaller size. The observed accuracies were as follows:
```  
  Data size  training accuracy  validation accuracy         model
0      3680           0.927717             0.938043       modelXY
1      3680           0.614946             0.593478      modelXY2
2       218           0.954128             0.750000  modelXYSmall
```
This phenomenon suggests that the dimensionality of the dataset (i.e., the number of features) plays a crucial role in model development. For instance, the shape of X2_train was (3680, 2) used for modelXY2, and the shape of X_small_train was (218, 57) used for modelXYSmall. The higher accuracy associated with X_small_train could be attributed to its greater number of features, indicating that a rich feature set may compensate for a smaller data size to some extent.

However, it's important to recognize that not all features contribute equally to model performance. The relevance of each feature to the target variable and the weighting of these features can significantly influence the accuracy of the model. This observation aligns with our discussions during lectures, where we emphasized the importance of both the quantity and quality of data in building effective models.

</b></font>

# **Part 4: Reflection (2 marks)**
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


<font color='Green'><b>
In this assignment, I enjoyed practicing with machine learning models and applying concepts we have learned in lecture, especially the ideas of splitting the data into a training and test set.
Something I found very interesting was the load_spam() dataset provided by yellow_bricks, since it never occured to me how emails could have been classified as spam or not with machine learning techniques.
</b></font>

# **Part 5: Bonus Question (4 marks)**

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [13]:
# import the libraries for regression
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_absolute_error, mean_squared_error

results = pd.DataFrame(columns=["Model", "alphaVal", "MSE", "MAE", "R2"])
maxAlphaRidgeR2 = 0
maxAlphaLassoR2 = 0
maxAlphaRidge = 0
maxAlphaLasso = 0
for alphaVal in range(0, 100, 1):
    # Create a Ridge Regression model
    model = Ridge(alpha=alphaVal)
    model.fit(X_train, y_train)

    ridge_predictions = model.predict(X_test)
    ridge_mse = mean_squared_error(y_test, ridge_predictions)
    ridge_mae = mean_absolute_error(y_test, ridge_predictions)
    ridge_r2 = r2_score(y_test, ridge_predictions)

    # Create a Lasso Regression model
    model = Lasso(alpha=alphaVal)
    model.fit(X_train, y_train)

    lasso_predictions = model.predict(X_test)
    lasso_mse = mean_squared_error(y_test, lasso_predictions)
    lasso_mae = mean_absolute_error(y_test, lasso_predictions)
    lasso_r2 = r2_score(y_test, lasso_predictions)

    # Add data to the pandas dataframe
    new_row = {"MSE": ridge_mse, "MAE": ridge_mae, "R2":ridge_r2, "Model": "Ridge", "alphaVal": alphaVal}
    new_row_df = pd.DataFrame([new_row])
    results = pd.concat([results, new_row_df])

    new_row = {"MSE": lasso_mse, "MAE": lasso_mae, "R2":lasso_r2, "Model": "Lasso", "alphaVal": alphaVal}
    new_row_df = pd.DataFrame([new_row])
    results = pd.concat([results, new_row_df])
    
    if ridge_r2 > maxAlphaRidgeR2:
        maxAlphaRidgeR2 = ridge_r2
        maxAlphaRidge = alphaVal
    if lasso_r2 > maxAlphaLassoR2:
        maxAlphaLassoR2 = lasso_r2
        maxAlphaLasso = alphaVal
print(results)
# print max R2 value of ridge and lasso
print(f"maxAlphaRidgeR2: {maxAlphaRidgeR2} alpha: {maxAlphaRidge}, maxAlphaLassoR2: {maxAlphaLassoR2} alpha: {maxAlphaLasso}")

c:\Users\night\miniconda3\Lib\site-packages\sklearn\base.py:1351: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  return fit_method(estimator, *args, **kwargs)
c:\Users\night\miniconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:678: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent(
c:\Users\night\miniconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.546e+04, tolerance: 2.326e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV instead.
  model = cd_fast.enet_coordi

    Model alphaVal         MSE       MAE        R2
0   Ridge        0   95.635335  7.865299  0.636898
0   Lasso        0   95.635335  7.865299  0.636898
0   Ridge        1   95.635231  7.865291  0.636899
0   Lasso        1   95.335850  7.840309  0.638035
0   Ridge        2   95.635128  7.865283  0.636899
..    ...      ...         ...       ...       ...
0   Lasso       97  124.090490  9.065582  0.528861
0   Ridge       98   95.625373  7.864555  0.636936
0   Lasso       98  124.766648  9.095662  0.526294
0   Ridge       99   95.625273  7.864547  0.636936
0   Lasso       99  125.450674  9.125743  0.523697

[200 rows x 5 columns]
maxAlphaRidgeR2: 0.6369363120550218 alpha: 99, maxAlphaLassoR2: 0.6388744784541781 alpha: 10


<font color='Green'><b>
The greatest R2 values and the alpha value for Lasso and Ridge are:
maxAlphaRidgeR2: 0.6369363120550218 alpha: 99
maxAlphaLassoR2: 0.6388744784541781 alpha: 10

From the data, we can see that the Lasso model with an alpha value of 10 performs slightly better than the Ridge model with the alpha value of 99.

The R^2 metric provides a measure of how well the observed outcomes are replicated by the model, based on the proportion of total variation of outcomes explained by the model. A higher R^2 value (close to 1) indicates a better fit to the data. Since the R^2 score is a score from 0 to 1, a score of 0.639 is a moderately high score.

Whether this score is "good enough" depends on the engineering application, context, and similar benchmarks.

</b></font>